### PR TITLE
Fixed broken tests from pull #18.  …

### PR DESCRIPTION
**Pull request by rosskevin**
I believe these assertions to be correct and I have been using the behavior asserted by this code for some time.

Peer review is welcome to see if this is the right change.

### CODE REVIEW

**Comment by rosskevin** [comment]
@mcgain mentions in #25

>@rosskevin you should think twice before changing that test. Rails' behaviour is that save returns nil when a transaction is rolled back.

**Comment by mcgain** [comment]
Pasting from other PR
 Rails' behaviour is that save returns nil when a transaction is rolled back.
```ruby
class User < ActiveRecord::Base
  after_save do
    raise ActiveRecord::Rollback
  end
end

class UserTest < ActiveSupport::TestCase
  test "ActiveRecord::Base#save returns nil on a rolled-back transaction" do
    assert_nil User.new(name: "Rupert").save
  end
end
```
Perhaps you're fine with breaking that behaviour, it is undocumented as far as I can tell.

File: test/machine_with_event_attributes_on_save_test.rb [hunk 1]
```diff
@@ -176,9 +176,9 @@ def test_should_allow_additional_transitions_to_previous_state_in_after_transiti
     assert_equal 'first_gear', @record.state
   end
 
-  def test_should_return_nil_on_manual_rollback
+  def test_should_return_false_on_before_transition_rollback
     @machine.before_transition { raise ActiveRecord::Rollback }
 
-    assert_equal nil, @record.save
+    assert_equal false, @record.save
```
Comment (rosskevin) [inline test/machine_with_event_attributes_on_save_test.rb]: This line is questionable based on @mcgain's comment about activerecord default behavior.

**Comment by mcgain** [comment]
My only thought about how to retain activerecord's behaviour would be to set `@success` to `nil` in `StateMachines::TransitionCollection` when a callback raises.

**Comment by rosskevin** [comment]
@mcgain I'm having a hard time seeing how [18's changes](https://github.com/state-machines/state_machines-activerecord/pull/18/files) affected this.  [`#around_save` had to be changed](https://github.com/state-machines/state_machines-activerecord/pull/18/files#diff-89e78a7be3b59d9fef66e24230de2158L510) in order to allow the state_machine to decide whether or not to use transactions.  Perhaps that yields the different result that I need to compensate for related to [the corresponding change in the parent project](https://github.com/state-machines/state_machines/pull/17/files)?

**Comment by rosskevin** [comment]
I'll look at that now.

**Comment by rosskevin** [comment]
In the parent project, specifically `TransitionCollection#run_actions`, it uses `catch_exceptions` block which would be a good place to intercept and introduce `@success = nil`. 

```ruby
      # The results of the actions will be used to determine #success?.
      def run_actions
        catch_exceptions do  #######################
          @success = if block_given?
            result = yield
            actions.each {|action| results[action] = result}
            !!result
          else
            actions.compact.each {|action| !skip_actions && results[action] = object.send(action)}
            results.values.all?
          end
        end
      end
       ...
      # Wraps the given block with a rescue handler so that any exceptions that
      # occur will automatically result in the transition rolling back any changes
      # that were made to the object involved.
      def catch_exceptions
        begin
          yield
        rescue Exception
          rollback
          raise
        end
      end
```

Unfortunately, this test is failing in the callback, and it is run without that block:

```ruby
      def run_callbacks(index = 0, &block)
        if transition = self[index]
          throw :halt unless transition.run_callbacks(:after => !skip_after) do
            run_callbacks(index + 1, &block)
            {:result => results[transition.action], :success => success?}
          end
        else
          persist
          run_actions(&block)
        end
      end
```
where `:halt` is not thrown (it is an ActiveRecord::Rollback in this case) and it escapes this block entirely.

Further, both `run_actions` and `run_callbacks` are run from `TransitionCollection#perform`, note the different transaction and error handling treatment of `run_actions` from `run_callbacks`:

```ruby
    def perform(&block)
      reset
      
      if valid?
        if use_event_attributes? && !block_given?
          each do |transition|
            transition.transient = true
            transition.machine.write(object, :event_transition, transition)
          end
          
          run_actions
        else
          within_transaction do
            catch(:halt) { run_callbacks(&block) }
            rollback unless success?
          end
        end
      end
      
      if actions.length == 1 && results.include?(actions.first)
        results[actions.first]
      else
        success?
      end
    end
```

I'm adding a few more tests that are really basic, but are adding some clarity for me, but so far it seems like TransitionCollection needs a bit of refactoring.

**Comment by rosskevin** [comment]
I'm wondering if we have more problems with the activerecord integration, or if this is related.  For example, this is failing with a validation error (though I cannot figure out where):

```ruby
  def test_should_yield_three_models_with_save!
    @machine.before_transition { @model.create!; puts "before ran, now #{@model.count}" }
    @machine.around_transition { @model.create!; puts "around ran, now #{@model.count}" }

    assert_equal true, @record.save!
    assert_equal 3, @model.count
  end
```
The execution is halted in [callback.rb](https://github.com/state-machines/state_machines/blob/master/lib/state_machines/callback.rb#L186).  

```
before ran, now 1
around ran, now 2

Minitest::UnexpectedError: ActiveRecord::RecordInvalid: Validation failed: 
    /Users/kross/state_machines-activerecord/lib/state_machines/integrations/active_record.rb:497:in `save!'
    /Users/kross/state_machines-activerecord/test/machine_with_event_attributes_on_save_test.rb:197:in `test_should_yield_three_models_with_save!'
/Users/kross/state_machines-activerecord/lib/state_machines/integrations/active_record.rb:497:in `save!'
/Users/kross/state_machines-activerecord/test/machine_with_event_attributes_on_save_test.rb:197:in `test_should_yield_three_models_with_save!'
```

**Comment by rosskevin** [comment]
^^^ well evidently adding the around_transition makes the above case fail.  Changing assertion to 2 and commenting that out succeeds.  I've got other things to work on, but perhaps I'm missing why that would fail.

**Comment by rosskevin** [comment]
Ok, @seuros I need [state_machines#19](https://github.com/state-machines/state_machines/pull/19) to be merged before this will pass.  Once that is, let me know, I can bump the version and travis can confirm.

We should now be mimicking the behavior of ActiveRecord.save on rollback (returns nil)

**Comment by rosskevin** [comment]
@seuros - I see you [release state_machines 0.4.0](https://rubygems.org/gems/state_machines/versions/0.4.0), thank you.  I would like to revisit how we do dependency restrictions, because now I'll need you to bump state_machines-activemodel.

This gem requires `'state_machines-activemodel', '~> 0.2.0'` and transitively `state_machines`.   `state_machines-activemodel` has SM minor locked at `state_machines ~> 0.3.0`.

This is the second time we've run into this in the few pull requests I've submitted.  I recommend we loosen up our SM family of version requirements to `>=`, especially in the activemodel project which is a dependency between this and the parent project.

So, either way, I'll still need a build/release with a new SM dependency in `state_machines-activemodel` (preferably `'state_machines', '>= 0.4.0'`) before this pull can update a dependency version and the build will pass.

**Comment by seuros** [comment]
Sound good to me. I will do that.

**Comment by rosskevin** [comment]
@seuros good to merge, then we can see if #25 passes before another release.

**Pull request merged**